### ISSUE TITLE
W1b: `PinholeCamera.scale` — the principal-point rescale rule contradicts every pixel grid in the library

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4263]
Four public entry points rescale a `PinholeCamera`'s principal point as `cx' = s·cx` — `PinholeCamera.scale`, `PinholeCamera.scale_`, `scale_pinhole` and `kornia.sensors.camera.PinholeModel.scale`. That is the half-pixel (COLMAP / "pixel corner at the origin") convention. Every pixel grid in kornia uses the *other* convention: `create_meshgrid(normalized_coordinates=False)` enumerates integer pixel **centres** `0 … W-1`, and `depth_to_3d`, `depth_to_3d_v2`, `pixel2cam`, `unproject_meshgrid` and `DepthWarper.grid` all build on it. Under integer pixel centres the correct rescale is `cx' = s·cx + (s−1)/2`.

So a user who resizes an image by `s` and rescales their intrinsics with any of these four functions gets a principal point that is off by `(s−1)/2` pixels relative to the grid the rest of the library will use to unproject it. At `s = 0.5, cx = 4` the four sites give `2.0`; the grid-consistent value is `1.75`.

The four sites agree with each other, so this is one convention, not a divergence — the question is whether it is the right one for this library.

## Repro

```python
import torch
from kornia.geometry.camera import PinholeCamera
from kornia.geometry.camera.pinhole import scale_pinhole
from kornia.sensors.camera import CameraModel, CameraModelType
from kornia.image import ImageSize
from kornia.utils import create_meshgrid

K = torch.eye(4)[None]
K[0, 0, 0] = K[0, 1, 1] = 100.0
K[0, 0, 2], K[0, 1, 2] = 4.0, 3.0          # cx = 4, cy = 3
E = torch.eye(4)[None]
cam = PinholeCamera(K.clone(), E.clone(), torch.tensor([6]), torch.tensor([8]))
s = 0.5

print("grid first/last pixel of a 2x3 grid :",
      create_meshgrid(2, 3, normalized_coordinates=False)[0, 0, 0].tolist(),
      create_meshgrid(2, 3, normalized_coordinates=False)[0, -1, -1].tolist())
print("PinholeCamera.scale(0.5).cx        :", cam.scale(torch.tensor([s])).cx.item())
c2 = PinholeCamera(K.clone(), E.clone(), torch.tensor([6.0]), torch.tensor([8.0]))
c2.scale_(s)
print("PinholeCamera.scale_(0.5).cx       :", c2.cx.item())
vec = torch.tensor([[100.0, 100.0, 4.0, 3.0, 6.0, 8.0, 0, 0, 0, 0, 0, 0]])
print("scale_pinhole(0.5)[0][:4]          :", scale_pinhole(vec, torch.tensor([s]))[0, :4].tolist())
pm = CameraModel(ImageSize(6, 8), CameraModelType.PINHOLE, torch.tensor([100.0, 100.0, 4.0, 3.0]))
print("PinholeModel.scale(0.5).params     :", pm.scale(torch.tensor(s)).params.tolist())
print("integer-pixel-centre rule s*cx+(s-1)/2:", s * 4.0 + (s - 1) / 2)
```

## Observed

```
grid first/last pixel of a 2x3 grid : [0.0, 0.0] [2.0, 1.0]
PinholeCamera.scale(0.5).cx        : 2.0
PinholeCamera.scale_(0.5).cx       : 2.0
scale_pinhole(0.5)[0][:4]          : [50.0, 50.0, 2.0, 1.5]
PinholeModel.scale(0.5).params     : [50.0, 50.0, 2.0, 1.5]
integer-pixel-centre rule s*cx+(s-1)/2: 1.75
```

## Expected

Two coherent outcomes, and the choice is a behavior change for every existing caller:

1. **Adopt the integer-pixel-centre rule** (`cx' = s·cx + (s−1)/2`) at all four sites, matching `create_meshgrid` and every unprojection path in the library. This is the self-consistent option, and it silently changes the numbers every current caller gets.
2. **Keep `cx' = s·cx`** and document, on every one of the four symbols and on the conventions page, that kornia's camera rescale uses the half-pixel convention while its grids use integer centres, so the two must not be mixed.

Whichever is chosen, it applies to all four sites at once. The sweep for other affected paths was executed and is empty: `grep -rn 'intrinsic|camera_matrix' kornia/augmentation kornia/geometry/transform` returns nothing, so no augmentation or resize path rescales intrinsics today. The repo-wide list of principal-point rescale sites is exactly `kornia/geometry/camera/pinhole.py`, `kornia/sensors/camera/camera_model.py`, `kornia/geometry/calibration/distort.py` and `kornia/geometry/epipolar/projection.py`.

## Classification

**Semantic contract** — one rescale rule for intrinsics, a different pixel-centre convention for every grid. Idiom 3 (a guard/contract that does not match the documented behavior, here across two modules rather than inside one function).

## Disposition

**MERGE IN WINDOW — behavior change.** Canonical: whichever rule the window picks, applied to all four sites at once (`PinholeCamera.scale`, `PinholeCamera.scale_`, `scale_pinhole`, `PinholeModel.scale`). This is the behavior-change discussion CONTRIBUTING asks for, and the decision is: **in the coordinated Repair & Deprecation window, not before. A PR that changes the rescale rule ahead of it will be closed and pointed here.** Batch 5 documents the current rule on all four symbols and pins it.

## Environment

kornia 0.9.0rc1 @ `92e5fe8d` (main), torch 2.14.0, python 3.11, cpu. Found during the camera conventions audit (batch 5, tracking issue #4262); the batch documents the behavior below as it is and pins it, and does not change it.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Comment by ducha-aiki** [comment]
Second observation on the same method, found while pinning this issue for batch 5 (audit label `5d-sc-38`): `kornia.sensors.camera.PinholeModel.scale` also changes the **type** of the `ImageSize` it returns. It rebuilds the size as `ImageSize(self.image_size.height * scale_factor, self.image_size.width * scale_factor)`, so the python `int` height and width the constructor accepted come back as 0-dim floating tensors on the model's device — and as `3.0` / `4.0` rather than the integer pixel counts an `ImageSize` is meant to hold.

```python
import torch
from kornia.sensors.camera import CameraModel, CameraModelType
from kornia.image import ImageSize

cam = CameraModel(ImageSize(6, 8), CameraModelType.PINHOLE, torch.tensor([100.0, 50.0, 4.0, 3.0]))
print("before:", cam.image_size, type(cam.image_size.height).__name__)
print("after :", cam.scale(torch.tensor(0.5)).image_size, type(cam.scale(torch.tensor(0.5)).image_size.height).__name__)
```

```
before: ImageSize(height=6, width=8) int
after : ImageSize(height=tensor(3.), width=tensor(4.)) Tensor
```

On `mps` the two fields carry `device='mps:0'` as well.

This is not the principal-point rule this issue is about, and it survives either of the two outcomes in the Expected section above — both of them change *what* `cx'` is, neither of them decides what `ImageSize` holds. Recording it here rather than as a separate issue, because whoever repairs `scale` touches these two lines together: **the repair has to decide whether `ImageSize` keeps python ints (rounding, and choosing a rounding rule) or is allowed to hold tensors.** `PinholeCamera.scale`, the other site with an image size, has the same question in the `height` / `width` form (`#4265` covers only the in-place `int64` crash there).

Batch 5d pins the current behaviour as `tests/sensors/camera/test_camera_model.py::TestPinholeCamera::test_wart_scale_turns_the_image_size_fields_into_tensors_4263`, which names this issue and is deleted with it; it asserts the `int` inputs, the `torch.Tensor` outputs, their 0-dim shape, their device and the values `3.0` / `4.0`.

Executed 2026-09-06 on kornia 0.9.0rc1 @ `e8111a61` (main), torch 2.14.0, python 3.11, cpu and mps.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Comment by ducha-aiki** [comment]
Two more findings for this issue's scope, from the batch-8 conventions audit (two-view geometry). Both are in `kornia/geometry/epipolar/projection.py`, which is already in the site list above.

**1. The rescale leaves the skew `K[0, 1]` unscaled.** `scale_intrinsics`, `PinholeCamera.scale` and `PinholeCamera.scale_` multiply `fx`, `fy`, `cx`, `cy` by `s` and leave `K[0, 1]` as it is. Resizing an image by `s` maps pixel coordinates through `S = [[s, 0, a], [0, s, a], [0, 0, 1]]`, with `a = 0` for the half-pixel rule and `a = (s - 1) / 2` for the integer-pixel-centre rule. The rescaled intrinsics are `S @ K`, and under both rules that multiplies the skew by `s`. So whichever outcome is chosen, the repair should scale `K[0, 1]` too. `scale_pinhole` and `PinholeModel.scale` have no skew parameter, so this concerns only the three sites that take a full `K`.

**2. `intrinsics_like` builds its principal point with the half-pixel rule.** It sets `(cx, cy) = (W / 2, H / 2)`, which is the image centre when pixel corners sit at integer coordinates. On kornia's integer-pixel-centre grid the centre of a `W x H` image is `((W - 1) / 2, (H - 1) / 2)`. This is the same convention split, at a site that constructs `K` instead of rescaling it. Under outcome 1 it becomes `((W - 1) / 2, (H - 1) / 2)`. Under outcome 2 its docstring has to say it uses the half-pixel centre. Either way it belongs to the same all-at-once decision, so the site list gains `intrinsics_like`.

```python
import torch
from kornia.geometry import create_meshgrid
from kornia.geometry.camera import PinholeCamera
from kornia.geometry.epipolar import intrinsics_like, scale_intrinsics

# (1) the skew K[0, 1] is not rescaled
K = torch.tensor([[[800.0, 3.0, 320.0], [0.0, 760.0, 200.0], [0.0, 0.0, 1.0]]])  # skew = 3
s = 2.0
K4 = torch.eye(4)[None]
K4[:, :3, :3] = K
cam = PinholeCamera(K4, torch.eye(4)[None], torch.tensor([480.0]), torch.tensor([640.0]))
print("scale_intrinsics(K, 2)[0, 1]           :", scale_intrinsics(K, s)[0, 0, 1].item())
print("PinholeCamera.scale(2) intrinsics[0, 1]:", cam.scale(torch.tensor([s])).intrinsics[0, 0, 1].item())
cam.scale_(s)
print("PinholeCamera.scale_(2) intrinsics[0,1]:", cam.intrinsics[0, 0, 1].item())
print("diag(s, s, 1) @ K [0, 1]               :", (torch.diag(torch.tensor([s, s, 1.0])) @ K[0])[0, 1].item())

# (2) intrinsics_like puts the principal point at (W / 2, H / 2)
img = torch.zeros(1, 1, 4, 6)  # H = 4, W = 6
Ki = intrinsics_like(500.0, img)
grid = create_meshgrid(4, 6, normalized_coordinates=False)[0]
print("intrinsics_like (cx, cy)               :", Ki[0, 0, 2].item(), Ki[0, 1, 2].item())
print("create_meshgrid pixel-centre mean (x,y):", grid[..., 0].mean().item(), grid[..., 1].mean().item())
```

```
scale_intrinsics(K, 2)[0, 1]           : 3.0
PinholeCamera.scale(2) intrinsics[0, 1]: 3.0
PinholeCamera.scale_(2) intrinsics[0,1]: 3.0
diag(s, s, 1) @ K [0, 1]               : 6.0
intrinsics_like (cx, cy)               : 3.0 2.0
create_meshgrid pixel-centre mean (x,y): 2.5 1.5
```

`scale_intrinsics` is bitwise equal to `PinholeCamera.scale(s).intrinsics[:, :3, :3]` and to `scale_pinhole` on the shared `fx, fy, cx, cy` (checked with a per-batch `s = (0.5, 3.0)` in float64), so `scale_intrinsics` follows the same rule as the four sites above and has to change with them. The batch-8 docs PR documents both behaviours as they are and links them here.

Executed 2026-09-24 on kornia main 7ddf731bd, torch 2.14.0 (and 2.5.1: same), CPU float32, macOS arm64.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).
